### DISCUSSION TITLE
Type mismatch between LSP WorkspaceEdit and vscode.WorkspaceEdit

### DISCUSSION THREAD

**Discussion by Greglar4** [microsoft/vscode-discussions#2209] [Extension Development QnA]
Hello,

I am working on an extension that uses a language server. I am using the [apply edit](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_applyEdit) request to modify resources on the client side. This requires sending a [WorkspaceEdit](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspaceEdit).

However, on the vscode client side, the [vscode.workspace.applyEdit](https://code.visualstudio.com/api/references/vscode-api#workspace) function takes an argument of the type [vscode.WorkspaceEdit](https://code.visualstudio.com/api/references/vscode-api#WorkspaceEdit), which is not the same type as the WorkspaceEdit sent by the language server, which forces me to construct a vscode.WorkspaceEdit object from the LSP WorkspaceEdit object before applying the edit.

Why is this? Is there some kind of util function that would do this construction for me?

Thanks,

Greg

**Comment by burkeholland** [comment]
I don't believe that there is a built-in utility for this. Manual construction would be the way to go.

**Comment by Greglar4** [comment]
Is there a reason why the vscode.WorkspaceEdit type does not comply with the LSP WorkspaceEdit type?

**Reply by robconery** [reply]
VS Code’s WorkspaceEdit differs from LSP’s because it’s an API for extensions, not a data structure for RPC. Key differences: LSP uses plain object mappings, string URIs, and pure data, while VS Code has mutation methods, vscode.Uri, and extra file operations. Converting requires manual mapping. I'm not an expert here so if I have this wrong please let me know!
